src/system/emscripten/network.c

```diff
@@ -0,0 +1,166 @@
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#include <arpa/inet.h>
+#include <emscripten/websocket.h>
+#include <netdb.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "zenoh-pico/collections/string.h"
+#include "zenoh-pico/config.h"
+#include "zenoh-pico/system/platform.h"
+#include "zenoh-pico/utils/logging.h"
+#include "zenoh-pico/utils/pointers.h"
+
+#if Z_LINK_WS == 1
+
+/*------------------ TCP sockets ------------------*/
+int8_t _z_create_endpoint_ws(_z_sys_net_endpoint_t *ep, const char *s_addr, const char *s_port) {
+    int8_t ret = _Z_RES_OK;
+
+    struct addrinfo hints;
+    (void)memset(&hints, 0, sizeof(hints));
+    hints.ai_family = PF_UNSPEC;  // Allow IPv4 or IPv6
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_flags = 0;
+    hints.ai_protocol = IPPROTO_TCP;
+
+    if (getaddrinfo(s_addr, s_port, &hints, &ep->_ipws) < 0) {
+        ret = _Z_ERR_GENERIC;
+    }
+
+    return ret;
+}
+
+void _z_free_endpoint_ws(_z_sys_net_endpoint_t *ep) { freeaddrinfo(ep->_ipws); }
+
+/*------------------ TCP sockets ------------------*/
+int8_t _z_open_ws(_z_sys_net_socket_t *sock, const _z_sys_net_endpoint_t rep, uint32_t tout) {
+    int8_t ret = _Z_RES_OK;
+
+    sock->_fd = socket(rep._ipws->ai_family, rep._ipws->ai_socktype, rep._ipws->ai_protocol);
+    if (sock->_fd != -1) {
+        // WARNING: commented because setsockopt is not implemented in emscripten
+        // if ((ret == _Z_RES_OK) && (setsockopt(sock->_fd, SOL_SOCKET, SO_RCVTIMEO, (char *)&tv, sizeof(tv)) < 0)) {
+        //     ret = _Z_ERR_GENERIC;
+        // }
+
+        struct addrinfo *it = NULL;
+        for (it = rep._ipws; it != NULL; it = it->ai_next) {
+            if ((ret == _Z_RES_OK) && (connect(sock->_fd, it->ai_addr, it->ai_addrlen) < 0)) {
+                break;
+                // WARNING: breaking here because connect returns -1 even if the
+                // underlying socket is actually open.
+
+                // if (it->ai_next == NULL) {
+                //     ret = _Z_ERR_GENERIC;
+                //     break;
+                // }
+            } else {
+                break;
+            }
+        }
+
+        // WARNING: workaround as connect returns before the websocket is
+        // actually open.
+        z_sleep_ms(100);
+
+        if (ret != _Z_RES_OK) {
+            close(sock->_fd);
+        }
+    } else {
+        ret = _Z_ERR_GENERIC;
+    }
+
+    return ret;
+}
+
+int8_t _z_listen_ws(_z_sys_net_socket_t *sock, const _z_sys_net_endpoint_t lep) {
+    int8_t ret = _Z_RES_OK;
+    (void)sock;
+    (void)lep;
+
+    // @TODO: To be implemented
+    ret = _Z_ERR_GENERIC;
+
+    return ret;
+}
+
+void _z_close_ws(_z_sys_net_socket_t *sock) { close(sock->_fd); }
+
+size_t _z_read_ws(const _z_sys_net_socket_t sock, uint8_t *ptr, size_t len) {
+    // WARNING: workaroud as the socket is non blocking. Try 500 times
```

Comment:
Instead of trying 500 times, it should try to read while the configured timeout does not expire. Since, `setsockopt` is not supported in emscripten, it should be manually done here.